dashboard/src/actions/authActions.js

```diff
@@ -2,114 +2,48 @@ import * as APP_ROUTES from "utils/routeConstants";
+ */
+export function waitForKeycloak(getState) {
+  const waitStart = Date.now();
+  const maxWait = 50000; // Milliseconds
+  /**
+   * Loop to check if the endpoints are loaded.
+   * @param {resolve} resolve object.
+   * @param {reject} reject object
+   */
+  function check(resolve, reject) {
+    if (Object.keys(getState().apiEndpoint.endpoints).length !== 0) {
+      resolve("Endpoints loaded");
+    } else if (Date.now() - waitStart > maxWait) {
+      reject(new Error("Timed out waiting for endpoints request"));
+    } else {
+      setTimeout(check, 250, resolve, reject);
     }
-};
-
-export const makeLoginRequest =
-  (details, navigate) => async (dispatch, getState) => {
-    try {
-      dispatch({ type: TYPES.LOADING });
-      // empty the alerts
-      dispatch({
-        type: TYPES.USER_NOTION_ALERTS,
-        payload: [],
-      });
-      const endpoints = getState().apiEndpoint.endpoints;
-      const response = await API.post(endpoints?.api?.login, {
-        ...details,
-      });
-      if (response.status === 200 && Object.keys(response.data).length > 0) {
-        const keepUser = getState().userAuth.keepLoggedIn;
-        const expiryTime = keepUser
-          ? CONSTANTS.EXPIRY_KEEPUSER_DAYS
-          : CONSTANTS.EXPIRY_DEFAULT_DAYS;
-        Cookies.set("isLoggedIn", true, { expires: expiryTime });
-        Cookies.set("token", response.data?.auth_token, {
-          expires: expiryTime,
-        });
-        Cookies.set("username", response.data?.username, {
-          expires: expiryTime,
-        });
-        const loginDetails = {
-          isLoggedIn: true,
-          token: response.data?.auth_token,
-          username: response.data?.username,
-        };
-        await dispatch({
-          type: TYPES.SET_LOGIN_DETAILS,
-          payload: loginDetails,
-        });
-
-        navigate(APP_ROUTES.OVERVIEW);
+  }
+  return new Promise((resolve, reject) => {
+    check(resolve, reject);
+  });
+}
 
-        dispatch(showToast(SUCCESS, "Logged in successfully!"));
-      }
-      dispatch({ type: TYPES.COMPLETED });
-    } catch (error) {
-      const alerts = getState().userAuth.alerts;
-      let alert = {};
-      if (error?.response) {
-        alert = {
-          title: error?.response?.data?.message,
-          key: uid(),
-        };
-        dispatch(toggleLoginBtn(true));
-      } else {
-        alert = {
-          title: error?.message,
-          key: uid(),
-        };
-        dispatch({ type: TYPES.NETWORK_ERROR });
-      }
-      alerts.push(alert);
-      dispatch({
-        type: TYPES.USER_NOTION_ALERTS,
-        payload: alerts,
-      });
-      dispatch({ type: TYPES.COMPLETED });
-    }
-  };
+// Perform some house keeping when the user logs in
+export const authCookies = () => async (dispatch, getState) => {
+  await waitForKeycloak(getState);
+  const keycloak = getState().apiEndpoint.keycloak;
+  if (keycloak.authenticated) {
+    Cookies.set("isLoggedIn", true, {
```

Comment:
This `if` statement has no `else` block...which is odd....

What happens if the user fails his/her authentication?  Are they "stuck" on the OIDC login page and never come back to the Dashboard?  Or, does OIDC somehow send them back with failure status?

That is, when execution reaches line 40 here, can `authenticated` ever be anything but `true`?  If so, shouldn't we have an `else` branch which acknowledges that the login failed (i.e, in opposition to line 44)?

Reply:
We could display `toastAction` like `else {dispatch(showToast("danger", "Not authenticated!"));}` but I don't think that's necessary. Landing on auth page without logging in, users already know that they are not logged in and we don't need to explicitly tell them that.

Reply:
I guess I'm confused about the sequencing.

When the Dashboard is initially loaded, we arrive at the code here _before_ the user does the Keycloak dance, but after we've initialized the Keycloak plug-in and created the `keycloak` object.  So, on the first pass through here, `keycloak.authenticated` will be false, and that's OK, and we'll go on to render some sort of helpful default page.  On subsequent page renderings (e.g., when the Keycloak login redirects back to the Dashboard), we'll come through here and `keycloak.authenticated` will be true, and so we'll set our cookie-thing here and raise a toast to the user's success.

So, to answer my own questions:  yes, we do want to test `keycloak.authenticated` here (because initially, at least, it **will** be false); no, we don't need an `else` clause on this `if` (because that execution path is "normal" and unremarkable); and, as you said, when an authentication attempt fails, we won't end up here.

But, execution will come through here at other times, right?  We don't necessarily want to toast the user every time, do we?  (We may or may not want to update the cookie...I'm less clear on that...but, then, I'm not sure that we need/want this cookie at all....)

So, I'm still wondering about this `if` statement:  perhaps it should be moved to somewhere where we know the answer and don't have to test for it.

Reply:
> I'm not sure that we need/want this cookie at all....

`isLoggedIn` cookie has an expiry of refreshToken so it lives across the multiple `App` rendering. I could check the value of `keycloak.authenticated` inside the protected route of the `App` but on each rendering, the keycloak object gets reinitialized and so the value is not there yet.  The problem is if I don't use this cookie in the protected route and instead rely on `keycloak.authenticated` value when the `App` renders it's gonna immediately display the DANGER toast and redirect the auth page. 

I also think ideally we shouldn't need to maintain cookies internally related to the authentication but I couldn't find a way around it at this time. This is something I need to spend some more time understanding the keycloak-js plugin and fix this reload problem.


> So, I'm still wondering about this `if` statement: perhaps it should be moved to somewhere where we know the answer and don't have to test for it.

What do you mean? We know the answer to what happens on the `if` statement and what will happen outside the `if` statement, right?

Reply:
> What do you mean?

We need the `if` statement here (we think...) because, at this point in the code, we don't know whether `keycloak.authenticated` is true or false.  The question is, is there some point, up- or down-stream from here, where we already know the answer?  If so, perhaps we can move the `if` statement there, so that we don't have to ask the question here.